@ariakit/test: simulated key events report code === "undefined" and option-less input events report inputType === "undefined"

### Current behavior

Every event built by `dispatch.<eventName>()` from `@ariakit/test` runs through `initEvent`, which overrides selected properties with own-property getters via `assignProps` ([dispatch.ts#L62-L69](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/dispatch.ts#L62-L69)), shadowing whatever the native event constructor set — including the spec defaults for fields the caller didn't pass.

`initKeyboardEvent` stringifies `key` and `code` unconditionally ([dispatch.ts#L145-L156](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/dispatch.ts#L145-L156)):

```ts
assignProps(event, {
  key: String(key),
  code: String(code),
  // ...
});
```

When the init dict omits them, `String(undefined)` evaluates to the literal string `"undefined"` — and since `"undefined"` is not nullish, the getter's `?? null` fallback never rescues it. `press()` and `type()` only ever pass `key`, never `code` ([press.ts#L228](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/press.ts#L228), [press.ts#L254](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/press.ts#L254), [type.ts#L71](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/type.ts#L71), [type.ts#L146](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/type.ts#L146)), so **every** simulated `keydown`/`keyup` reports `event.code === "undefined"`. Likewise, a bare `dispatch.keyDown(element)` reports `event.key === "undefined"`.

`initInputEvent` has the same flaw for `inputType` ([dispatch.ts#L84-L93](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/dispatch.ts#L84-L93)). Two helpers fire `input` events with no options at all and therefore produce `event.inputType === "undefined"`: `incrementNumberInput`, used by ArrowUp/ArrowDown on number inputs ([press.ts#L61-L69](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/press.ts#L61-L69)), and the `<option>` click path ([click.ts#L121-L122](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/click.ts#L121-L122)).

Per the UI Events and Input Events specs, `KeyboardEvent.code` and `InputEvent.inputType` default to the empty string `""` — which is what the native constructors in real browsers, jsdom, and happy-dom all set. So `"undefined"` is a truthy garbage string where every conforming environment reports a falsy `""` (or, for real typing, a physical code such as `"KeyM"`). Any handler under test that branches on `event.code` — the modern replacement for the deprecated `keyCode`, recommended for layout-independent shortcuts — or on `if (event.inputType)` behaves differently under `@ariakit/test` than under a real browser: a code-based shortcut map never matches, and `inputType` truthiness checks take the wrong branch.

The stringify pattern was inherited from `@testing-library/user-event` (see the provenance comment at [dispatch.ts#L1](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/dispatch.ts#L1)); upstream never hits this case because user-event always supplies both `key` and `code` from its keyboard map, so the unconditional `String(...)` here is an oversight rather than intent.

The same user scenario was verified in real Chromium with Playwright: pressing M delivers `code === "KeyM"` (the shortcut fires) and ArrowUp's `input` event never carries the string `"undefined"` — in fact Chromium fires a plain `Event` without any `inputType` for number-input spin-stepping.

### Steps to reproduce the bug

For a testing library, running a test is the consumer reproduction. In the Ariakit repo:

1. Create `app/src/sandbox/test-undefined-key-code/index.react.tsx` — a media-player-style widget whose keyboard shortcuts match on the physical `event.code`, plus a number input whose handler reads `InputEvent.inputType`; both values are mirrored into `<output>` elements (`JSON.stringify` keeps `""` and `"undefined"` distinguishable):

```tsx
import type { FormEvent, KeyboardEvent as ReactKeyboardEvent } from "react";
import { useState } from "react";

const shortcuts: Record<string, string> = {
  KeyM: "Mute",
  Space: "Play/Pause",
};

function display(value: string | null) {
  if (value === null) return "none";
  return JSON.stringify(value);
}

export default function Example() {
  const [lastKey, setLastKey] = useState<string | null>(null);
  const [lastCode, setLastCode] = useState<string | null>(null);
  const [action, setAction] = useState("none");
  const [lastInputType, setLastInputType] = useState<string | null>(null);

  function handleKeyDown(event: ReactKeyboardEvent<HTMLButtonElement>) {
    setLastKey(event.key);
    setLastCode(event.code);
    setAction(shortcuts[event.code] ?? "none");
  }

  function handleInput(event: FormEvent<HTMLInputElement>) {
    const { nativeEvent } = event;
    if (nativeEvent instanceof InputEvent) {
      setLastInputType(nativeEvent.inputType);
    } else {
      setLastInputType("(not an InputEvent)");
    }
  }

  return (
    <div>
      <button type="button" onKeyDown={handleKeyDown}>
        Player
      </button>
      <label htmlFor="volume">Volume</label>
      <input
        id="volume"
        type="number"
        defaultValue={5}
        min={0}
        max={10}
        onInput={handleInput}
      />
      <p>
        <output aria-label="Last key">{display(lastKey)}</output>{" "}
        <output aria-label="Last code">{display(lastCode)}</output>{" "}
        <output aria-label="Action">{action}</output>{" "}
        <output aria-label="Last input type">{display(lastInputType)}</output>
      </p>
    </div>
  );
}
```

2. Create `app/src/sandbox/test-undefined-key-code/test.ts`:

```ts
import { press, q } from "@ariakit/test";
import { expect, test } from "vitest";

function getOutputText(name: string) {
  return q.status.ensure(name).textContent;
}

test("keydown reports a spec-compliant code", async () => {
  await press("m", q.button.ensure("Player"));

  // Sanity check: the keydown was delivered with the right key.
  expect(getOutputText("Last key")).toBe('"m"');
  // Browsers report the physical code or the spec default ""; never the
  // literal string "undefined".
  expect(getOutputText("Last code")).not.toBe('"undefined"');
});

test("ArrowUp on a number input reports a spec-compliant inputType", async () => {
  const volume = q.spinbutton.ensure("Volume") as HTMLInputElement;
  await press.ArrowUp(volume);

  // Sanity check: the arrow key stepped the value and fired an input event.
  expect(volume.value).toBe("6");
  expect(getOutputText("Last input type")).not.toBe("none");
  // Browsers report a spec-listed inputType or the spec default ""; never
  // the literal string "undefined".
  expect(getOutputText("Last input type")).not.toBe('"undefined"');
});
```

3. Run `pnpm test-react test-undefined-key-code` from the repo root. Both tests fail:

```
AssertionError: expected '"undefined"' not to be '"undefined"'   // Last code
AssertionError: expected '"undefined"' not to be '"undefined"'   // Last input type
```

The exact same user scenario was verified in a real browser with Playwright (`app/src/sandbox/test-undefined-key-code/test-browser.ts`, real key presses on the rendered page): both tests pass in Chromium, which reports `code === "KeyM"` (so the shortcut fires) and never the string `"undefined"`.

### Expected behavior

Fields the caller doesn't pass should keep the value the native event constructor set — the spec default `""` for `KeyboardEvent.code`/`KeyboardEvent.key` and `InputEvent.inputType`. That's what real browsers, jsdom, and happy-dom all report for a constructor-built event with no `code`/`inputType` (happy-dom 20.10.2 probed directly: `new KeyboardEvent("keydown", { key: "m" }).code === ""`, `new InputEvent("input").inputType === ""`). The literal string `"undefined"` is a truthy garbage value no conforming environment ever produces, so `event.code`-based shortcut handling and `event.inputType` branching behave differently under `@ariakit/test` than everywhere else.

### Workaround

Consumer-side, in test code only.

For key events, `press`/`type` accept a `KeyboardEventInit` options argument that is forwarded into the keydown/keyup events, so pass the physical code explicitly:

```ts
await press("m", q.button.ensure("Player"), { code: "KeyM" });
```

For the `input` events fired by number-input arrow stepping, `press.ArrowUp()` offers no option passthrough, so replicate the step with the exported lower-level helpers, passing an explicit `inputType`:

```ts
import { dispatch, focus } from "@ariakit/test";

const volume = q.spinbutton.ensure("Volume") as HTMLInputElement;
await focus(volume);
await dispatch.keyDown(volume, { key: "ArrowUp", code: "ArrowUp" });
volume.value = (volume.valueAsNumber + 1).toString();
await dispatch.input(volume, { inputType: "" });
await dispatch.change(volume);
await dispatch.keyUp(volume, { key: "ArrowUp", code: "ArrowUp" });
```

Verified: with these in place, both tests above pass on the current `@ariakit/test` (and the code-based shortcut fires, since the event now carries a real `"KeyM"`).

### Possible solutions

Don't stringify absent values — fall back to the spec default `""` before converting, so the getter overrides reproduce native constructor behavior (`""` is not nullish, so the `value ?? null` in `assignProps` correctly yields `""` rather than `null`):

```diff
 function initInputEvent(
   event: InputEvent,
   { data, inputType, isComposing }: InputEventInit,
 ) {
   assignProps(event, {
     data,
     isComposing: !!isComposing,
-    inputType: String(inputType),
+    inputType: String(inputType ?? ""),
   });
 }
```

```diff
 function initKeyboardEvent(
   event: KeyboardEvent,
   { key, code, location, repeat, isComposing }: KeyboardEventInit,
 ) {
   assignProps(event, {
-    key: String(key),
-    code: String(code),
+    key: String(key ?? ""),
+    code: String(code ?? ""),
     location: sanitizeNumber(location),
     repeat: !!repeat,
     isComposing: !!isComposing,
   });
 }
```

The `key ?? ""` change also fixes the bare `dispatch.keyDown(element)` case. Both diffs were applied locally and validated: the repro tests above pass, and the `@ariakit/test` package's own suites stay green (5 files, 33 tests). A more faithful follow-up would be for `press`/`type` to derive a `code` from the key the way user-event's keyboard map does (`"Enter"` → `"Enter"`, `" "` → `"Space"`, `"a"` → `"KeyA"`), but that requires a key-to-code table and is layout-dependent for printable characters; the minimal fix already restores parity with what conforming DOMs report for a constructor-built event. Since `@ariakit/test` is a published package and this changes observable event values, a patch changeset is warranted.

*Found by an automated multi-agent audit of the repo at commit [`11bd4458d`](https://github.com/ariakit/ariakit/tree/11bd4458de6447b996dd882ea0c85eb801a6cd03) and verified with a real-browser Playwright reproduction (user-level interactions only).*

--- a/.changeset/300-test-event-defaults.md
+++ b/.changeset/300-test-event-defaults.md
@@ -1,0 +1,5 @@
+---
+"@ariakit/test": patch
+---
+
+Fixed `@ariakit/test` event helpers to use empty-string defaults instead of `"undefined"` for omitted keyboard and input event string fields.

--- a/packages/ariakit-test/src/dispatch.test.ts
+++ b/packages/ariakit-test/src/dispatch.test.ts
@@ -1,0 +1,93 @@
+import { expect, test } from "vitest";
+import { dispatch, press } from "./index.ts";
+
+test("dispatch.keyDown uses empty strings for omitted keyboard strings", async () => {
+  const button = document.createElement("button");
+  document.body.append(button);
+  let key: string | undefined;
+  let code: string | undefined;
+  button.addEventListener("keydown", (event) => {
+    key = event.key;
+    code = event.code;
+  });
+  try {
+    await dispatch.keyDown(button);
+    expect(key).toBe("");
+    expect(code).toBe("");
+  } finally {
+    button.remove();
+  }
+});
+
+test("dispatch.keyDown preserves provided keyboard strings", async () => {
+  const button = document.createElement("button");
+  document.body.append(button);
+  let key: string | undefined;
+  let code: string | undefined;
+  button.addEventListener("keydown", (event) => {
+    key = event.key;
+    code = event.code;
+  });
+  try {
+    await dispatch.keyDown(button, { key: "m", code: "KeyM" });
+    expect(key).toBe("m");
+    expect(code).toBe("KeyM");
+  } finally {
+    button.remove();
+  }
+});
+
+test("press uses an empty string for omitted keyboard code", async () => {
+  const button = document.createElement("button");
+  document.body.append(button);
+  let key: string | undefined;
+  let code: string | undefined;
+  button.addEventListener("keydown", (event) => {
+    key = event.key;
+    code = event.code;
+  });
+  try {
+    await press("m", button);
+    expect(key).toBe("m");
+    expect(code).toBe("");
+  } finally {
+    button.remove();
+  }
+});
+
+test("press.ArrowUp uses an empty inputType when stepping a number input", async () => {
+  const input = document.createElement("input");
+  input.type = "number";
+  input.value = "5";
+  document.body.append(input);
+  let inputType: string | undefined;
+  input.addEventListener("input", (event) => {
+    if (event instanceof InputEvent) {
+      inputType = event.inputType;
+    }
+  });
+  try {
+    await press.ArrowUp(input);
+    expect(input.value).toBe("6");
+    expect(inputType).toBe("");
+  } finally {
+    input.remove();
+  }
+});
+
+test("dispatch.input preserves provided inputType", async () => {
+  const input = document.createElement("input");
+  document.body.append(input);
+  let inputType: string | undefined;
+  input.addEventListener("input", (event) => {
+    if (event instanceof InputEvent) {
+      inputType = event.inputType;
+    }
+  });
+  try {
+    await dispatch.input(input, { inputType: "insertReplacementText" });
+    expect(inputType).toBe("insertReplacementText");
+  } finally {
+    input.remove();
+  }
+});

--- a/packages/ariakit-test/src/dispatch.ts
+++ b/packages/ariakit-test/src/dispatch.ts
@@ -72,6 +72,10 @@ function sanitizeNumber(n: number | undefined) {
   return n ?? 0;
 }
 
+function sanitizeString(value: string | undefined) {
+  return value ?? "";
+}
+
 function initClipboardEvent(
   event: ClipboardEvent,
   { clipboardData }: ClipboardEventInit,
@@ -88,7 +92,7 @@ function initInputEvent(
   assignProps(event, {
     data,
     isComposing: !!isComposing,
-    inputType: String(inputType),
+    inputType: sanitizeString(inputType),
   });
 }
 
@@ -147,8 +151,8 @@ function initKeyboardEvent(
   { key, code, location, repeat, isComposing, charCode }: KeyboardEventInit,
 ) {
   assignProps(event, {
-    key: String(key),
-    code: String(code),
+    key: sanitizeString(key),
+    code: sanitizeString(code),
     location: sanitizeNumber(location),
     repeat: !!repeat,
     isComposing: !!isComposing,

